inference_overview.md

```diff
@@ -0,0 +1,449 @@
+TEE, since it uses private features. These models will tend to be larger. Both
+these factors can increase the cost and latency of running such models.
+
+### Factorized models
+
+**Model factorization** is a technique that makes it possible to break a single
+model into multiple pieces, and then combine those pieces into a prediction. In
+B&A-supported use cases, models often make use of three kinds of data: user
+data, contextual data, and ad data.
+
+As described above, in the unified case, a single model is trained on all three
+kinds of data. In the factorized case, you could break the model up into
+multiple pieces. The model that includes the sensitive user data needs to be
+executed within the trust boundary, on the buyer's bidding services.
+
+This makes the following design possible:
+
+1. Break the model up into a private piece (the sensitive user data) and one or
+  more non-private pieces (the contextual and ad data).
+1. Optionally, inferences from some or all of the non-private pieces can be
+  passed in as arguments to a [User Defined Function][1] (UDF), e.g.
+  `generateBid()`. For example, contextual embeddings can be computed along the
+  contextual path, returned to the seller's frontend in the contextual response,
+  and then can be passed into the B&A auction as part of the `buyer_signals`
+  constructed by the seller when they initiate the Auction.
+1. Optionally, ad techs can create models for non-private pieces (such as ads
+  features) ahead of time, then materialize embeddings from those models into
+  the ad retrieval server and fetch these embeddings at runtime.
+1. To make a prediction during a UDF, combine private embeddings from the
+  inference service running in a TEE with non-private embeddings from UDF
+  function arguments or the ad retrieval server store with an operation like a
+  dot product. This is the final prediction.
+
+The following figure shows a factorized pCTR model broken into three parts,
+contextual, user and ads towers. Only the User tower needs to run inside the
+TEE. The contextual and ad towers can run outside the TEE since they do not rely
+on private features.
+
+<figure id = "image-1">
+  <img src = "images/inference_overview_fig1.png"
+  width = "100%"
+  alt = "Factorized pCTR model">
+  <figcaption><b>Figure 1.</b> Factorized pCTR Model </figcaption>
+</figure>
+
+Factorized models are useful because loading and executing large models inside
+the TEE may be expensive. Oftentimes, only a small part of the model relies on
+private data (like user data), and the rest needs non-private (e.g. contextual
+or ad) data. With factorization, parts that depend on non-private features can
+be outside the TEE where they can be executed on ad tech hardware, while keeping
+the models requiring private features inside the TEE.
+
+Factorization can be utilized only when the model can be broken down into its
+constituents (as described above) fairly easily. In case models rely on a lot of
```

Comment:
For clarity: If models depend excessively on cross features (e.g., ad x contextual information), the factorization may not be an effective strategy.

Reply:
Question. Why is it not a good strategy in this case?

Reply:
Done